packages/firestore/src/util/async_queue.ts

```diff
@@ -223,17 +261,55 @@ export class AsyncQueue {
   }
 
   /**
-   * Waits until all currently scheduled tasks are finished executing. Tasks
-   * scheduled with a delay can be rejected or queued for immediate execution.
+   * Waits until all currently queued tasks are finished executing. Delayed
+   * operations are not run.
    */
-  drain(executeDelayedTasks: boolean): Promise<void> {
-    this.delayedOperations.forEach(delayedOp => {
-      if (executeDelayedTasks) {
-        delayedOp.skipDelay();
-      } else {
-        delayedOp.cancel('shutdown');
+  drain(): Promise<void> {
+    return this.enqueue(() => Promise.resolve());
+  }
+
+  /**
+   * For Tests: Determine if a particular delayed operation exists.
+   */
+  hasDelayedOperation(timerId: TimerId): boolean {
+    return this.delayedOperations.findIndex(op => op.timerId === timerId) >= 0;
+  }
+
+  /**
+   * For Tests: Runs some or all delayed operations early.
+   *
+   * @param lastTimerId If specified, only delayed operations up to and
+   *   including this TimerId will be drained. Throws if no such operation
+   *   exists.
+   * @returns a Promise that resolves once all operations have been run.
+   */
+  runDelayedOperationsEarly(lastTimerId?: TimerId): Promise<void> {
+    // Note that draining may generate more delayed ops, so we do that first.
+    return this.drain().then(() => {
+      assert(
+        lastTimerId === undefined || this.hasDelayedOperation(lastTimerId),
+        `Attempted to drain to missing operation ${lastTimerId}`
+      );
+
+      // Run ops in the same order they'd run if they ran naturally.
+      this.delayedOperations.sort((a, b) => a.targetTime - b.targetTime);
+
```

Comment:
This seems like a lot of hoops to go through that (at least for now) I don't cleary see the immediate benefit of. For unrelated operations scheduled with delay, it seems fine to a) run them out of order (removing the need for target time) and b) only run them when calling `runDelayedOperationsEarly`.